router/src/cluster_based.rs

```diff
@@ -87,32 +90,47 @@ impl ClusterBasedRouter {
             .with_context(|| OtherWithCause {
                 msg: format!("Failed to route tables by cluster, req:{route_tables_req:?}"),
             })?;
+        trace!("Route tables by cluster, req:{route_tables_req:?}, resp:{route_resp:?}");
 
         // Now we pick up the nodes who own the leader shard for the route response.
         for (table_name, route_entry) in route_resp.entries {
-            for node_shard in route_entry.node_shards {
-                if node_shard.shard_info.is_leader() {
-                    let route = make_route(route_entry.table_info, &node_shard.endpoint)?;
-                    if let Some(cache) = &self.cache {
-                        // There may be data race here, and it is acceptable currently.
-                        cache.insert(table_name.clone(), route.clone()).await;
+            let route = if route_entry.node_shards.is_empty() {
+                Some(make_route(route_entry.table_info, None)?)
+            } else {
+                let mut route = None;
+                for node_shard in route_entry.node_shards {
+                    if node_shard.shard_info.is_leader() {
+                        route = Some(make_route(
+                            route_entry.table_info,
+                            Some(&node_shard.endpoint),
+                        )?);
+                        break;
                     }
-                    routes.push(route);
-                    break;
                 }
+                route
```

Comment:
Shall we use the find to make it simple: `route_entry.node_shards.into_iter().find(..)`?